pathways/inspections.py

```diff
@@ -234,29 +226,82 @@ def compute_max_inspectable_stems(num_stems, stems_per_box, within_box_pct):
 
-    # Select boxes to inspect
+    # Select units to inspect
     selection_strategy = config["inspection"]["selection_strategy"]
     if selection_strategy == "tailgate":
-        box_index_to_inspect = range(n_boxes_to_inspect)
+        index_to_inspect = range(n_units_to_inspect)
     elif selection_strategy == "random":
-        box_index_to_inspect = random.sample(range(num_boxes), n_boxes_to_inspect)
+        if unit in ["stem", "stems"]:
+            index_to_inspect = random.sample(range(num_stems), n_units_to_inspect)
+        elif unit in ["box", "boxes"]:
+            index_to_inspect = random.sample(range(num_boxes), n_units_to_inspect)
+        else:
+            raise RuntimeError("Unknown unit: {unit}".format(**locals()))
+    elif selection_strategy == "hierarchical":
+        outer = config["inspection"]["hierarchical"]["outer"]
+        if unit in ["stem", "stems"]:
+            n_boxes_to_inspect = convert_stems_to_boxes(
+                config, shipment, n_units_to_inspect
+            )
+            max_stems = compute_max_inspectable_stems(
+                num_stems, stems_per_box, within_box_pct
+            )
+            if outer == "random":
+                if max_stems >= n_units_to_inspect:
+                    index_to_inspect = random.sample(
+                        range(num_boxes), n_boxes_to_inspect
+                    )
+                else:
+                    raise RuntimeError(
+                        "Num stems to inspect ({n_units_to_inspect}) is greater "
+                        "than max inspectable stems ({max_stems}). Increase "
+                        "within_box_pct.".format(**locals())
+                    )
+            elif outer == "interval":
+                interval = config["inspection"]["hierarchical"]["interval"]
+                max_stems = math.floor(max_stems / interval)
+                if max_stems >= n_units_to_inspect:
+                    index_to_inspect = []
+                    index = 0
+                    for i in range(n_boxes_to_inspect):
+                        index_to_inspect.append(index)
+                        index += interval
+                else:
+                    raise RuntimeError(
+                        "Num stems to inspect ({n_units_to_inspect}) is greater "
+                        "than max inspectable stems ({max_stems}). Increase "
+                        "within_box_pct or decrease interval.".format(**locals())
+                    )
+            else:
+                raise RuntimeError("Unknown outer unit: {outer}".format(**locals()))
+        elif unit in ["box", "boxes"]:
+            raise RuntimeError(
+                "Cannot use hierarchical selection strategy with box sampling unit"
+            )
+        else:
+            raise RuntimeError("Unknown unit: {unit}".format(**locals()))
     else:
         raise RuntimeError(
             "Unknown selection strategy: {selection_strategy}".format(**locals())
```

Comment:
If you put this whole block into a separate function (which would be called here), it could make the inspect() function more readable and will reduce some of the Pylint complains (or the need to disable them).